File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2017/Y17D16.kt
package org.gristle.pdxpuzzles.advent.y2017

import org.gristle.pdxpuzzles.advent.utilities.Day

class Y17D16(input: String) : Day {

    sealed interface DanceMove {
        @JvmInline
        value class Spin(val value: Int) : DanceMove
        class Exchange(val a: Int, val b: Int) : DanceMove
        class Partner(val a: Char, val b: Char) : DanceMove
    }

    // parsing input to DanceMoves
    private val danceMoves: List<DanceMove> by lazy {
        input
            .split(',')
            .map {
                when (it[0]) {
                    's' -> DanceMove.Spin(it.drop(1).toInt())
                    'x' -> it.drop(1).split('/').let { (a, b) -> DanceMove.Exchange(a.toInt(), b.toInt()) }
                    'p' -> it.drop(1).split('/').let { (a, b) -> DanceMove.Partner(a[0], b[0]) }
                    else -> throw IllegalArgumentException("Cannot parse input: $it")
                }
            }
    }

    // takes a string, runs all dance moves on it, and returns a string
    private fun danceParty(positions: String): String {

        // convert String to mutable primitive array for performance
        val pos = positions.toCharArray()

        // rather than reorder entire array after each spin, just keep track of the offset and reverse it at the end 
        var offset = 0

        // execute each instruction
        for (danceMove in danceMoves) {
            when (danceMove) {
                is DanceMove.Exchange -> {
                    val a = (danceMove.a - offset).mod(pos.size)
                    val b = (danceMove.b - offset).mod(pos.size)
                    pos[a] = pos[b].also { pos[b] = pos[a] }
                }

                is DanceMove.Partner -> {
                    val a = pos.indexOf(danceMove.a)
                    val b = pos.indexOf(danceMove.b)
                    pos[a] = pos[b].also { pos[b] = pos[a] }
                }

                is DanceMove.Spin -> offset = (danceMove.value + offset).mod(pos.size)
            }
        }

        // generate a string from the charArray, using the offset information to put the first character first, etc.
        return unShift(pos, offset)
    }

    private fun unShift(pos: CharArray, index: Int) = buildString {
        pos.indices.forEach { i -> append(pos[(i - index).mod(pos.size)]) }
    }

    private val start = "abcdefghijklmnop"

    // store results of first round so that it can be reused for part 2
    private val oneRound: String by lazy { danceParty(start) }

    override fun part1() = oneRound

    override fun part2(): String {

        // cache stores strings representing program arrangements that have already occurred  
        val cache = mutableSetOf(start)

        // run danceParty on its own output over and over again, storing each result to cache, until a duplicate is
        // found
        generateSequence(oneRound, ::danceParty)
            .first { !cache.add(it) }

        // use mod to calculate what would result if the dance party were run 1 billion times
        return cache.elementAt(1_000_000_000 % cache.size)
    }
}

fun main() = Day.runDay(Y17D16::class)

//    Class creation: 16ms
//    Part 1: hmefajngplkidocb (2ms)
//    Part 2: fbidepghmjklcnoa (35ms)
//    Total time: 53ms

File: rust/advent/src/bin/2017/y17d16.rs
use advent::utilities::get_input::get_input;
use indexmap::IndexSet;
use itertools::Itertools;
use utilities::structs::stopwatch::{ReportDuration, Stopwatch};

type Input = ([u8; 16], Vec<DanceMove>);
type Output = String;

#[derive(Debug, Clone, Copy)]
enum DanceMove {
    Spin(i32),
    Exchange(i32, i32),
    Partner(u8, u8),
}

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(17, 16).unwrap();
    let input = parse_input(&input);
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

fn parse_input(input: &str) -> Input {
    let positions: [u8; 16] = "abcdefghijklmnop".as_bytes().into_iter().copied()
        .collect_vec()
        .try_into()
        .unwrap();

    let dance_moves = input.split(',')
        .map(|it| {
            let (dance_code, args) = it.split_at(1);
            match dance_code {
                "s" => DanceMove::Spin(args.parse().unwrap()),
                "x" => {
                    let (a, b) = args.split('/').map(|it| it.parse().unwrap()).collect_tuple().unwrap();
                    DanceMove::Exchange(a, b)
                },
                "p" => {
                    let (a, b) = args.split('/')
                        .map(|it| it.chars().next().unwrap() as u8)
                        .collect_tuple()
                        .unwrap();
                    DanceMove::Partner(a, b)
                },
                _ => unreachable!(),
            }
        })
        .collect();

    (positions, dance_moves)
}

fn dance_party(positions: &mut [u8], dance_moves: &[DanceMove]) {
    // rather than reorder entire array after each spin, just keep track of the offset and reverse it at the end
    let mut offset = 0;
    let len = positions.len() as i32;

    // execute each instruction
    for &dance_move in dance_moves {
        match dance_move {
            DanceMove::Spin(spin) => {
                offset = (spin + offset) % len as i32;
            },
            DanceMove::Exchange(a, b) => {
                let a = (a - offset).rem_euclid(len) as usize;
                let b = (b - offset).rem_euclid(len) as usize;
                positions.swap(a, b);
            },
            DanceMove::Partner(a, b) => {
                let a = positions.iter().position(|&it| it == a).unwrap();
                let b = positions.iter().position(|&it| it == b).unwrap();
                positions.swap(a, b);
            },
        }
    }

    // rotate to accommodate the spin
    positions.rotate_right(offset as usize);
}

fn part1(input: &Input) -> Output {
    let (positions, dance_moves) = input;
    let mut positions = *positions;
    dance_party(&mut positions, dance_moves);
    positions.into_iter().map(|b| b as char).collect()
}

fn part2(input: &Input) -> Output {
    let (positions, dance_moves) = input;
    let mut positions = *positions;
    let mut cache: IndexSet<[u8; 16]> = IndexSet::new();
    loop {
        if !cache.insert(positions.clone()) {
            let finale = cache.get_index(1_000_000_000 % cache.len()).unwrap();
            return finale.iter().map(|&b| b as char).collect();
        }
        dance_party(&mut positions, dance_moves);
    }
}

#[test]
fn default() {
    let input = get_input(17, 16).unwrap();
    let input = parse_input(&input);
    assert_eq!("hmefajngplkidocb".to_string(), part1(&input));
    assert_eq!("fbidepghmjklcnoa".to_string(), part2(&input));
}

// Input parsed (333μs)
// 1. hmefajngplkidocb (56μs)
// 2. fbidepghmjklcnoa (2ms)
// Total: 2ms
